specs/CustomDownload.md

```diff
@@ -0,0 +1,771 @@
+                                                L"Mime type: " + mimeType.get() + L"\r\n" +
+                                                L"Expected Download size in bytes: " +
+                                                std::to_wstring(expectedDownloadSizeInBytes) +
+                                                L"\r\n";
+
+                    TextInputDialog dialog(
+                        m_appWindow->GetMainWindow(), L"Download Starting", prompt.c_str(),
+                        description.c_str(), resultFilePath.get());
+                    if (dialog.confirmed)
+                    {
+                        // If user selects `OK`, the download will complete normally.
+                        // Result file path will be updated if a new one was provided.
+                        CHECK_FAILURE(args->put_ResultFilePath(dialog.input.c_str()));
+                        UpdateProgress(download.get());
+                    }
+                    else
+                    {
+                        // If user selects `Cancel`, the download will be canceled.
+                        CHECK_FAILURE(args->put_Cancel(TRUE));
+                    }
+                };
+
+                // Obtain a deferral for the event so that the CoreWebView2
+                // doesn't examine the properties we set on the event args until
+                // after we call the Complete method asynchronously later.
+                wil::com_ptr<ICoreWebView2Deferral> deferral;
+                CHECK_FAILURE(args->GetDeferral(&deferral));
+
+                // This function can be called to show the download dialog and
+                // complete the event at a later time, allowing the developer to
+                // perform async work before the event completes.
+                m_completeDeferredDownloadEvent = [showDialog, deferral] {
+                    showDialog();
+                    CHECK_FAILURE(deferral->Complete());
+                };
+
+                return S_OK;
+            })
+        .Get(),
+    &m_downloadStartingToken));
+}
+
+// Update download progress
+void ScenarioCustomDownloadExperience::UpdateProgress(ICoreWebView2Download* download)
+{
+    // Register a handler for the `DownloadProgressSizeInBytesChanged` event.
+    CHECK_FAILURE(download->add_DownloadProgressSizeInBytesChanged(
+        Callback<ICoreWebView2DownloadProgressSizeInBytesChangedEventHandler>(
+            [this](ICoreWebView2Download* download, IUnknown* args) -> HRESULT {
+                // Here developer can update UI to show progress of a download using
+                // `download->get_DownloadProgressSizeInBytes` and
+                // `download->get_ExpectedDownloadSizeInBytes`.
+                return S_OK;
+            })
+        .Get(),
+    &m_downloadProgressSizeInBytesChangedToken));
+
+    // Register a handler for the `DownloadStateChanged` event.
+    CHECK_FAILURE(download->add_DownloadStateChanged(
```

Comment:
`StateChanged` is sufficient since it's already on a `Download` object, which is the sender of the event.

Reply:
See comments in the IDL below.